### ISSUE TITLE
“option str” broken for ByteStrings

### ISSUE THREAD

**Issue by llelf** [pcapriotti/optparse-applicative#368]
Broken Haskell Hello World:

```sh
$ ./opt-app-bs --msg Мир
Hello, "\FS8@"
```

The fix: stop using `IsString` in optparse-applicative, it’s broken by design (because of nonsense instance `bytestring` provides).

```haskell
import Data.ByteString (ByteString)
import Options.Applicative

data Sample = Sample { msg :: ByteString }

sample :: Parser Sample
sample = Sample <$> strOption (long "msg")

main :: IO ()
main = execParser opts >>= \(Sample msg) -> putStrLn $ "Hello, " <> show msg

opts :: ParserInfo Sample
opts = info (sample <**> helper) (fullDesc <> header "Broken Hello World in Haskell")
```

**Comment by HuwCampbell** [comment]
I'm not sure how you would assume this should behave; bytes need encodings.

I adopted an `IsString` based `str` because it was the easiest, most portable way of giving people lazy and strict `Text` readers without depending on the text library itself. There was even a `optparse-text` package which provided a `ReadM Strict.Text` reader because we couldn't do this naturally!

I honestly don't believe that wobbly bytestring instance is worth removing this capability.

**Comment by llelf** [comment]
> how you would assume this should behave

it would behave in only one obvious behaviour, printing `Пример, Мир`

>  bytes need encodings.

Yes they do.  In the context of optparse-applicative one sane choice would be UTF-8.

In any way, **mangling text is wrong** and bug.

**Comment by HuwCampbell** [comment]
> Yet currently optparse-applicative somehow constructs bytes without it.

optparse isn't doing *anything* with regards to constructing a `Bytestring`, it doesn't even have a bytestring dependency.

> In any way, mangling text is wrong and bug.

I don't even disagree, it's just that this is a bytestring issue, not an optparse one.

```haskell
>> import Data.ByteString
>> :set -XOverloadedStrings 
>> show ("Мир" :: ByteString)
"\"\\FS8@\""
```

**Comment by llelf** [comment]
> optparse isn't doing anything with regards to constructing a Bytestring

of course it does. It is given a valid `args::[String]` and it produces invalid `ByteString`.

> it's just that this is a bytestring issue, not an optparse one

You cannot use IsString for non-ASCII bytestrings, this is part of Haskell folklore now unfortunately.

optparse-applicative is a very popular library, can we please not make Haskell Unicode situation any worse than it is?

**Comment by hasufell** [comment]
> I'm not sure how you would assume this should behave; bytes need encodings.

Well, not really. Bytes are just bytes.

On unix systems, getArgs returns you the bytes and you are free to interpret them (e.g. by utilizing the current locale or assuming something random like UTF-8).

> I don't even disagree, it's just that this is a bytestring issue, not an optparse one.

Well yes, IsString is kinda broken: https://github.com/haskell/bytestring/issues/140 and it probably won't be fixed anytime soon. The solution is to not use IsString.

Silently truncating chars for a cli parser is a pretty huge bug.

**Comment by HuwCampbell** [comment]
> Well yes, IsString is kinda broken: [haskell/bytestring#140](https://github.com/haskell/bytestring/issues/140) and it probably won't be fixed anytime soon. The solution is to not use IsString.

Two points: that thread does seem pretty active, so hopefully we'll see some movement there; in the meantime, almost every Haskell code base today does use IsString though, in the form of overloaded strings, so I wouldn't say that "don't use is" is always tenable.

I would say that my communication and the documentation has been lacking here though; adding a note to the documentation for `str` saying something like:

> Using bytestring instances of `IsString` is not recommended. These instance encode with ISO-8859-1 and will silently truncate characters beyond the Latin-1 supplemental block. The functions in `Data.Text.Encoding` can be used if other encodings are desired.
> ```haskell
> Data.Text.Encoding.encodeUtf8 <$> str
> ```

could be helpful.

**Comment by hasufell** [comment]
Semi-related to this: I wonder how much work it would be to allow getting the args unchanged as ByteString on unix? Would that require a fork?

**Comment by HuwCampbell** [comment]
This is addressed somewhat in #65.
